### PR TITLE
remove glib dependency by providing compatible replacements

### PR DESCRIPTION
**Pull request by cseagle**
Added glib compatible replacements for all glib functions used in unicorn. Removed glib dependencies in Makefiles.
I threw the list and hash table implementations together and have paid no attention to their performance. Feel free to replace the hash table implementation with something that performs better.
Need help testing. All samples and unit tests run with same results as master on the following platforms:
win32 (using mingw), OS X 64, Linux 32, Linux 64.
Have not done much testing beyond that.

### CODE REVIEW

**Comment by aquynh** [comment]
very nice! did you write this glib compatible code from scratch?

**Comment by cseagle** [comment]
Yes, it's all from scratch.

**Comment by cseagle** [comment]
I did use a couple of public hash algorithms.

**Comment by aquynh** [comment]
could you point out the source?

what do you think about just reusing glibc code instead, so we dont need to review it?
looks like the related code in glibc is self-contain, so we can just copy it, with some simple modification to make it work for us.

**Comment by cseagle** [comment]
The source of one hash is already noted as djb2. It's essentially the same hash that glib uses as it turns out with one minor change. The integer hash is from Bob Jenkins here: http://burtleburtle.net/bob/hash/integer.html.

If you want to copy from glib, that's fine. It will be more code with more dependencies that either need to be taken in their entirety or trimmed down. It will include a lot of code that is never used by unicorn/qemu. Someone will need to work out all of the inter-dependencies with the glib bits. For example, their hash table implementation and perhaps other things too make use of a custom memory allocator (gslice) that would need to be pulled in as well.

File: qemu/glib_compat.c [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+   return hash;
+}
+
+int g_int_equal(const void *v1, const void *v2) {
+   return *(const int*)v1 == *(const int*)v2;
+}
+
+/* Doubly-linked list */
+
+GList *g_list_first(GList *list) {
```
Comment (danghvu) [inline qemu/glib_compat.c]: function style
```
{
}
```

File: qemu/glib_compat.c [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+GList *g_list_prepend(GList *list, void* data) {
+   GList *n = (GList*)g_malloc(sizeof(GList));
+   n->next = list;
+   n->prev = NULL;
+   n->data = data;
+   return n;
+}
+
+GList *g_list_remove_link(GList *list, GList *llink) {
+   if (llink == list) list = list->next;
```
Comment (danghvu) [inline qemu/glib_compat.c]: `if (llink)` ?

**Comment by aquynh** [comment]
now this cannot compile on Linux/GCC: https://travis-ci.org/unicorn-engine/unicorn/jobs/185052000

**Comment by cseagle** [comment]
Your comment is useless. That's the prototype straight from glib. Take it up with them.

**Comment by cseagle** [comment]
It compiles just fine for me. clock_gettime is a linux system call, are you telling me it is not present on the travis build system?

**Comment by cseagle** [comment]
Feel free to change it.

**Comment by aquynh** [comment]
i merged this into a new branch "noglib" for now.

**Comment by lunixbochs** [comment]
Does QEMU build using _GNU_SOURCE defined? Some of the libc headers are gated under that, and glib would definitely have it defined.

**Comment by aquynh** [comment]
https://travis-ci.org/unicorn-engine/unicorn/jobs/185133337: for some weird reason, Travis cannot find clock_gettime(), while it did before this patch. @stephengroat, do you have any ideas?

File: qemu/target-arm/helper.c [hunk 3]
```diff
@@ -155,7 +155,7 @@ static void count_cpreg(gpointer key, gpointer opaque)
     }
 }
 
-static gint cpreg_key_compare(gconstpointer a, gconstpointer b)
+static int32_t cpreg_key_compare(gconstpointer a, gconstpointer b)
```
Comment (aquynh) [inline qemu/target-arm/helper.c]: i think `gint` is `int`, but not int32_t ? see https://developer.gnome.org/glib/stable/glib-Basic-Types.html#gint

**Comment by stephengroat** [comment]
try adding `-lrt` to the `samples` Makefile library list and see if that helps. was probably included through glib before

**Comment by stephengroat** [comment]
same yourself some build time, replace your travis file in your branch with this:

````
language: c
sudo: false
before_install:
  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew update; fi
  - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then brew install cmocka; fi
  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ./install-cmocka-linux.sh; fi
script:
  - make && make -C bindings/go && make -C bindings/go test && make test
compiler:
  - clang
  - gcc
os:
  - linux
  - osx
addons:
  apt:
    packages:
    # mingw-w64 packages too old in precise
````

File: qemu/include/glib_compat.h [hunk 4]
```diff
@@ -104,7 +103,7 @@ void g_hash_table_remove_all(GHashTable *hash_table);
 int g_hash_table_remove(GHashTable *hash_table, const void* key);
 void g_hash_table_unref(GHashTable *hash_table);
 GHashTable *g_hash_table_ref(GHashTable *hash_table);
-uint32_t g_hash_table_size(GHashTable *hash_table);
+guint g_hash_table_size(GHashTable *hash_table);
```
Comment (radare) [inline qemu/include/glib_compat.h]: guint is not always 32 bit

File: docs/COMPILE-NIX.md [hunk 5]
```diff
@@ -8,24 +8,10 @@ To compile for Microsoft Windows, see [COMPILE-WINDOWS.md](COMPILE-WINDOWS.md)
...
-  64-bit), you need to have glib in universal format. To do that, build &
-  install glib with:
-
-        $ brew install glib --universal
-
-- For Linux, "glib2-dev" is needed.
-  Ubuntu/Debian users can install this with:
-
-        $ sudo apt-get install libglib2.0-dev
+For MacOS, Unicorn requires "pkg-config" package.
```
Comment (radare) [inline docs/COMPILE-NIX.md]: Isnt pkgconfig required only for glib?

Comment (aquynh) [inline docs/COMPILE-NIX.md]: pkgconfig is checked in qemu/configure, but looks like we dont need that anymore. i will remove this dependency.

**Comment by aquynh** [comment]
@stephengroat, now `MACOS_UNIVERSAL=yes` is default, but you suggested to remove that `brew install glib --universal cmocka` in Travis setup, why?

**Comment by aquynh** [comment]
@cseagle, i merged your latest commits to "noglib" branch now.

**Comment by aquynh** [comment]
@stephengroat, adding -lrt does not help: https://travis-ci.org/unicorn-engine/unicorn/jobs/185361656.

looks like this only happens with older gcc. i dont have this issue on Ubuntu 16.04

**Comment by stephengroat** [comment]
@aquynh you only use `LDFLAGS` in your compile step since you reversed my default compile changes 😄 

add the `-lrt` to the `LDFLAGS` line and you'll have more luck!

**Comment by aquynh** [comment]
adding -lrt to LDFLAGS causes error on MacOS, like below:

```
  LINK    sample_arm
ld: library not found for -lrt
```

you are right, i just removed glib from Travis & Appveryor setup.

File: qemu/glib_compat.c [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "glib_compat.h"
+
+#ifdef _WIN32
+   #ifdef _WIN64
+      #define __HAVE_64_BIT_PTRS
```
Comment (aquynh) [inline qemu/glib_compat.c]: do we need these macros __HAVE_64_BIT_PTRS & __HAVE_32_BIT_PTRS? i cant find where they are used

Comment (aquynh) [inline qemu/glib_compat.c]: oops sorry, i missed its use in the hash function below.

**Comment by aquynh** [comment]
now Travis can build Linux samples by adding `-lrt` to link flags: https://github.com/unicorn-engine/unicorn/commit/b34109fb0d64ba4b6b3f45f96afd391f81553999.

@lunixbochs, but it is Go's turn now, can you take a look: https://travis-ci.org/unicorn-engine/unicorn/jobs/186505305 ? i think Go also needs `-lrt`, but not sure where to put that.

**Comment by lunixbochs** [comment]
LDFLAGS in a cgo comment. -lunicorn is in there somewhere. but it might need a linux specific tag? I don't know which platforms need it.

**Comment by aquynh** [comment]
do you mean something like this https://github.com/unicorn-engine/unicorn/commit/af4c29c57fcc60c2fc7ffec5c799d5b799a6a0fe ?

it seems only Linux needs this flag `-lrt`.

**Comment by lunixbochs** [comment]
Exactly that, but with a linux tag. I think like this: "#cgo linux LDFLAGS"

While we're at it that include is wrong. uc.h is the private header right? It should probably use <unicorn/unicorn.h> and fix that -I../ to be /include

**Comment by aquynh** [comment]
to be sure (so that i will not make a mistake), can you send a PR here?

**Comment by lunixbochs** [comment]
not in a good place to do that

**Comment by aquynh** [comment]
@cseagle, i made some minor fixes to make the glib_compat API better compatible with the original glib API: https://github.com/unicorn-engine/unicorn/commit/6b57da1fc2122376bb0c9439c2908987b62c9a15

**Comment by cseagle** [comment]
Ok thanks.

**Comment by aquynh** [comment]
i am not really sure about the performance of this new glib_compat code, so i lifted some of the most complicated functions regarding hash, sort & string from Glib to replace the original code, and push to the new branch at https://github.com/unicorn-engine/unicorn/tree/noglib2. the rest is still from the code of @cseagle. please help to test this new branch.

**Comment by lunixbochs** [comment]
What's the glib hash table function? Could be worth replacing it with something like murmur or city or whatever google is using these days while you're in there.

**Comment by aquynh** [comment]
at least we are on the same level as before because we are still using glib code inside. i am not really sure replacing proven stable code with new one would improve anything here.

**Comment by cseagle** [comment]
Did you do any concrete testing or was this "just because"

**Comment by cseagle** [comment]
Make sure you pull in all the relevant copyright notices, the glib code has a number of different contributors and copyright notices

**Comment by aquynh** [comment]
@cseagle no, except running our existing tests in Travis & Appveyor setup. i copied as-is code from Glib, so (hopefully) they are bug-free.

on the license, i will copy it over from glib. thanks for reminding.

**Comment by cseagle** [comment]
The other issue this brings up is tracking upstream for bug fixes. If a bug is fixed in this portion of the glib code will you be tracking that for porting?

**Comment by aquynh** [comment]
yes i will port the bugfixes over.

**Comment by aquynh** [comment]
i merged this PR & the branch "noglib2" into "master", thanks!

**Pull request closed without merging**